src/Optimizely/Event/Builder/EventBuilder.php

```diff
@@ -216,14 +242,15 @@ private function setConversionParams($config, $eventKey, $experimentVariationMap
      */
     public function createImpressionEvent($config, $experimentKey, $variationKey, $userId, $attributes)
     {
-        $this->resetParams();
-        $this->setCommonParams($config, $userId, $attributes);
+        $eventParams = $this->getCommonParams($config, $userId, $attributes);
 
         $experiment = $config->getExperimentFromKey($experimentKey);
         $variation = $config->getVariationFromKey($experimentKey, $variationKey);
-        $this->setImpressionParams($experiment, $variation->getId());
+        $impressionParams = $this->getImpressionParams($experiment, $variation->getId());
+
+        $eventParams[VISITORS][0]['snapshots'][] = $impressionParams;
```

Comment:
If this logic can be in a function would be good, please check C# code.

Reply:
The commonParams and EventSpecificParam merging code is only one-line, and is slightly (See empty square brackets] different for both events.